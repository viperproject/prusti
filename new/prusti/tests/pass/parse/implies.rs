// compile-flags: -Zprint-desugared-specs -Zprint-typeckd-specs
// normalize-stdout-test: "[a-z0-9]{32}" -> "$(NUM_UUID)"
// normalize-stdout-test: "[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}" -> "$(UUID)"

use prusti_contracts::*;

// #[requires(forall(|a: Map<u32,u32>, b: u32| a == 5 ==> b == 10, triggers=[(1+1,), (2,3), (4,)]))]
// #[requires((true ==> true) || true)]
// #[requires(true && true || true)]
// #[requires((true && true) || true ==> forall(|| 1+1) || true && true)]
#[requires(true ==> true)]
fn test1() {}

// #[ensures((1+1 == 2) && ((1 + 1) == 2))]
// fn test2() {}
//
// fn test3() {
//     for _ in 0..2 {
//         invariant!(true)
//     }
// }
//
// #[requires(true)]
// #[ensures(true)]
// fn test4() {
//     for _ in 0..2 {
//         invariant!(true)
//     }
// }

fn main() {}
