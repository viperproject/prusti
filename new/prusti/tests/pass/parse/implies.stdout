// compile-flags: -Zprint-desugared-specs -Zprint-typeckd-specs -Zskip-verify -Zhide-uuids
// normalize-stdout-test: "[a-z0-9]{32}" -> "$(NUM_UUID)"
// normalize-stdout-test: "[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}" -> "$(UUID)"

#![feature(register_tool)]
#![register_tool(prusti)]
#[prelude_import]
use std::prelude::v1::*;
#[macro_use]
extern crate std;

use prusti_contracts::*;

#[prusti::spec_only]
#[prusti::spec_id = "$(NUM_UUID)"]
#[prusti::assertion =
  "{/"kind/":{/"Implies/":[{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":101}}},{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":102}}}]}}"]
fn prusti_pre_item_test1_$(NUM_UUID)() {

    #[prusti::expr_id = "$(NUM_UUID)_101"]
    || -> bool { true };

    #[prusti::expr_id = "$(NUM_UUID)_102"]
    || -> bool { true };
}
#[prusti::pre_spec_id_ref = "$(NUM_UUID)"]
fn test1() { }

#[prusti::spec_only]
#[prusti::spec_id = "$(NUM_UUID)"]
#[prusti::assertion =
  "{/"kind/":{/"Implies/":[{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":101}}},{/"kind/":{/"Implies/":[{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":102}}},{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":103}}}]}}]}}"]
fn prusti_pre_item_test2_$(NUM_UUID)() {

    #[prusti::expr_id = "$(NUM_UUID)_101"]
    || -> bool { true };

    #[prusti::expr_id = "$(NUM_UUID)_102"]
    || -> bool { true };

    #[prusti::expr_id = "$(NUM_UUID)_103"]
    || -> bool { true };
}
#[prusti::pre_spec_id_ref = "$(NUM_UUID)"]
fn test2() { }

#[prusti::spec_only]
#[prusti::spec_id = "$(NUM_UUID)"]
#[prusti::assertion =
  "{/"kind/":{/"Implies/":[{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":101}}},{/"kind/":{/"Implies/":[{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":102}}},{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":103}}}]}}]}}"]
fn prusti_pre_item_test3_$(NUM_UUID)() {

    #[prusti::expr_id = "$(NUM_UUID)_101"]
    || -> bool { true };

    #[prusti::expr_id = "$(NUM_UUID)_102"]
    || -> bool { true };

    #[prusti::expr_id = "$(NUM_UUID)_103"]
    || -> bool { true };
}
#[prusti::pre_spec_id_ref = "$(NUM_UUID)"]
fn test3() { }

#[prusti::spec_only]
#[prusti::spec_id = "$(NUM_UUID)"]
#[prusti::assertion =
  "{/"kind/":{/"Implies/":[{/"kind/":{/"Implies/":[{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":101}}},{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":102}}}]}},{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":103}}}]}}"]
fn prusti_pre_item_test4_$(NUM_UUID)() {

    #[prusti::expr_id = "$(NUM_UUID)_101"]
    || -> bool { true };

    #[prusti::expr_id = "$(NUM_UUID)_102"]
    || -> bool { true };

    #[prusti::expr_id = "$(NUM_UUID)_103"]
    || -> bool { true };
}
#[prusti::pre_spec_id_ref = "$(NUM_UUID)"]
fn test4() { }

#[prusti::spec_only]
#[prusti::spec_id = "$(NUM_UUID)"]
#[prusti::assertion =
  "{/"kind/":{/"Implies/":[{/"kind/":{/"Implies/":[{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":101}}},{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":102}}}]}},{/"kind/":{/"Implies/":[{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":103}}},{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":104}}}]}}]}}"]
fn prusti_pre_item_test5_$(NUM_UUID)() {

    #[prusti::expr_id = "$(NUM_UUID)_101"]
    || -> bool { true };

    #[prusti::expr_id = "$(NUM_UUID)_102"]
    || -> bool { true };

    #[prusti::expr_id = "$(NUM_UUID)_103"]
    || -> bool { true };

    #[prusti::expr_id = "$(NUM_UUID)_104"]
    || -> bool { true };
}
#[prusti::pre_spec_id_ref = "$(NUM_UUID)"]
fn test5() { }

fn main() { }
Procedure(ProcedureSpecification { pres: [Assertion { kind: Implies(Assertion { kind: Expr(Expression { spec_id: SpecificationId($(UUID)), id: ExpressionId(101), expr: BodyId { hir_id: HirId { owner: DefId(0:5 ~ implies[317d]::prusti_pre_item_test1_$(NUM_UUID)[0]), local_id: 5 } } }) }, Assertion { kind: Expr(Expression { spec_id: SpecificationId($(UUID)), id: ExpressionId(102), expr: BodyId { hir_id: HirId { owner: DefId(0:5 ~ implies[317d]::prusti_pre_item_test1_$(NUM_UUID)[0]), local_id: 12 } } }) }) }], posts: [] })
Procedure(ProcedureSpecification { pres: [Assertion { kind: Implies(Assertion { kind: Expr(Expression { spec_id: SpecificationId($(UUID)), id: ExpressionId(101), expr: BodyId { hir_id: HirId { owner: DefId(0:9 ~ implies[317d]::prusti_pre_item_test2_$(NUM_UUID)[0]), local_id: 5 } } }) }, Assertion { kind: Implies(Assertion { kind: Expr(Expression { spec_id: SpecificationId($(UUID)), id: ExpressionId(102), expr: BodyId { hir_id: HirId { owner: DefId(0:9 ~ implies[317d]::prusti_pre_item_test2_$(NUM_UUID)[0]), local_id: 12 } } }) }, Assertion { kind: Expr(Expression { spec_id: SpecificationId($(UUID)), id: ExpressionId(103), expr: BodyId { hir_id: HirId { owner: DefId(0:9 ~ implies[317d]::prusti_pre_item_test2_$(NUM_UUID)[0]), local_id: 19 } } }) }) }) }], posts: [] })
Procedure(ProcedureSpecification { pres: [Assertion { kind: Implies(Assertion { kind: Expr(Expression { spec_id: SpecificationId($(UUID)), id: ExpressionId(101), expr: BodyId { hir_id: HirId { owner: DefId(0:14 ~ implies[317d]::prusti_pre_item_test3_$(NUM_UUID)[0]), local_id: 5 } } }) }, Assertion { kind: Implies(Assertion { kind: Expr(Expression { spec_id: SpecificationId($(UUID)), id: ExpressionId(102), expr: BodyId { hir_id: HirId { owner: DefId(0:14 ~ implies[317d]::prusti_pre_item_test3_$(NUM_UUID)[0]), local_id: 12 } } }) }, Assertion { kind: Expr(Expression { spec_id: SpecificationId($(UUID)), id: ExpressionId(103), expr: BodyId { hir_id: HirId { owner: DefId(0:14 ~ implies[317d]::prusti_pre_item_test3_$(NUM_UUID)[0]), local_id: 19 } } }) }) }) }], posts: [] })
Procedure(ProcedureSpecification { pres: [Assertion { kind: Implies(Assertion { kind: Implies(Assertion { kind: Expr(Expression { spec_id: SpecificationId($(UUID)), id: ExpressionId(101), expr: BodyId { hir_id: HirId { owner: DefId(0:19 ~ implies[317d]::prusti_pre_item_test4_$(NUM_UUID)[0]), local_id: 5 } } }) }, Assertion { kind: Expr(Expression { spec_id: SpecificationId($(UUID)), id: ExpressionId(102), expr: BodyId { hir_id: HirId { owner: DefId(0:19 ~ implies[317d]::prusti_pre_item_test4_$(NUM_UUID)[0]), local_id: 12 } } }) }) }, Assertion { kind: Expr(Expression { spec_id: SpecificationId($(UUID)), id: ExpressionId(103), expr: BodyId { hir_id: HirId { owner: DefId(0:19 ~ implies[317d]::prusti_pre_item_test4_$(NUM_UUID)[0]), local_id: 19 } } }) }) }], posts: [] })
Procedure(ProcedureSpecification { pres: [Assertion { kind: Implies(Assertion { kind: Implies(Assertion { kind: Expr(Expression { spec_id: SpecificationId($(UUID)), id: ExpressionId(101), expr: BodyId { hir_id: HirId { owner: DefId(0:24 ~ implies[317d]::prusti_pre_item_test5_$(NUM_UUID)[0]), local_id: 5 } } }) }, Assertion { kind: Expr(Expression { spec_id: SpecificationId($(UUID)), id: ExpressionId(102), expr: BodyId { hir_id: HirId { owner: DefId(0:24 ~ implies[317d]::prusti_pre_item_test5_$(NUM_UUID)[0]), local_id: 12 } } }) }) }, Assertion { kind: Implies(Assertion { kind: Expr(Expression { spec_id: SpecificationId($(UUID)), id: ExpressionId(103), expr: BodyId { hir_id: HirId { owner: DefId(0:24 ~ implies[317d]::prusti_pre_item_test5_$(NUM_UUID)[0]), local_id: 19 } } }) }, Assertion { kind: Expr(Expression { spec_id: SpecificationId($(UUID)), id: ExpressionId(104), expr: BodyId { hir_id: HirId { owner: DefId(0:24 ~ implies[317d]::prusti_pre_item_test5_$(NUM_UUID)[0]), local_id: 26 } } }) }) }) }], posts: [] })
